### PR TITLE
feat: add `LawfulMonad` instance for `EStateM`

### PR DESCRIPTION
**Pull request by eric-wieser**
Replaces the majority of leanprover-community/mathlib4#8772

### CODE REVIEW

**Comment by eric-wieser** [comment]
This is needed to provide a trivial reflexive `ULiftable IO IO` instance in mathlib.
Unfortunately that typeclass currently encodes lawfulness.

Even if IO is made opaque, it would be possible (and arguably a good idea) to make it so in a way that is still lawful.

Worst case scenario, we delete these instances if/when core changes.

**Comment by fgdorais** [comment]
> I'm a little skeptical here. We are planning on making the `IO` monad opaque in core, so you can't even see that it is defined in terms of `EStateM`.

I'm confused here. It seems obvious that `IO` and friends should be lawful, even if that turns out to be axiomatic. Of course, this is in core territory and shouldn't be in Std. There are no good circumstances where Std should add new axioms.

**Comment by eric-wieser** [comment]
Axioms are an irrelevant distraction here, there are axiom-free ways to make an opaque definition lawful (such as subtypes)

File: Std/Classes/LawfulMonad.lean [hunk 1]
```diff
@@ -52,6 +52,28 @@ instance : LawfulMonad Option := LawfulMonad.mk'
...
 
+
+instance : LawfulMonad (EStateM ε σ) := .mk'
+  (id_map := fun x => funext <| fun s => by
+    dsimp only [EStateM.instMonadEStateM, EStateM.map]
+    match x s with
+    | .ok _ _ => rfl
+    | .error _ _ => rfl)
+  (pure_bind := fun a x => rfl)
+  (bind_assoc := fun x f g => funext <| fun s => by
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Std/Classes/LawfulMonad.lean:63]: Does the CI not catch this? Adding the name here results in a nicer goal state when interacting with/scraping data from proofs

**Comment by digama0** [comment]
@eric-wieser 
> > I'm confused here. It seems obvious that `IO` and friends should be lawful, even if that turns out to be axiomatic. Of course, this is in core territory and shouldn't be in Std. There are no good circumstances where Std should add new axioms.
>
> Axioms are an irrelevant distraction here, there are axiom-free ways to make an opaque definition lawful (such as subtypes)

These two comments are connected. You can make an opaque definition lawful if you are the one defining it, but if the opaque definition is upstream then you need an `axiom` to make it lawful. `IO` is defined in core, so if it is defined opaquely then the lawfulness will need to be bundled with it or else Std would have to use an axiom to do it.

I share Scott's concern, I don't see the merit in making IO a LawfulMonad. In fact, I'm not sure I even agree that it should be lawful, this depends on the model we end up with for IO and I'm not sure there is a model that isn't a bag of lies so I think we should not encourage people to do proofs about IO functions and just make everything as opaque as possible. If mathlib wants to do this it should go in mathlib.

File: Std/Classes/LawfulMonad.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Std/Classes/LawfulMonad.lean:63]: @digama0, is that an argument that I should remove these names anyway, or that I should leave them be?

**Comment by joehendrix** [comment]
I concur there should be an instance of  `LawfulMonad` for `IO`.  If it's not true, then that reflects a bug in `IO` (or some subtle mistake in `LawfulMonad` that needs to be addressed).  

Given this [comment](https://github.com/leanprover/lean4/blob/14296ae720a97985eb35db35818d66c5ec75d63a/src/Init/System/IO.lean#L25) though, unless this is critical to address now, I think we should put in core.  I don't want to create additional work for making `IO`opaque.  

Does anyone know of blockers in Mathlib that would prevent `IO` from being opague?

**Comment by digama0** [comment]
> I would very much like this instance available in mathlib

So why are you PRing it here?

**Comment by eric-wieser** [comment]
Because all the other instances for `LawfulMonad` on builtin monad types are in Std, so I assumed it belonged next to them.

**Comment by digama0** [comment]
My recommendation is to keep the `EStateM` instance but move the rest to mathlib.

**Comment by eric-wieser** [comment]
I was just about to suggest the same compromise

**Comment by eric-wieser** [comment]
Alright, this now contains only the `EStateM` instance.

**Review by joehendrix — APPROVED** [review]

**Pull request merged**